File: tools/tf-types-ts/src/core/packet.ts
/**
 * Packet mode (TF-0011): standalone signed/encrypted objects that can
 * be delivered offline, relayed, stored, transferred, or verified
 * later. Every packet carries a priority class, expiration, route
 * constraints, and an ed25519 signature; large packets fragment into
 * an indexed sequence with a payload digest so any reassembled output
 * can be byte-checked against the original.
 *
 * Wire encoding is JSON canonicalisation by default; CBOR is offered
 * as a compact alternative for LoRa / serial transports. Compression
 * is optional (deflate). Both must be honoured by signers and
 * verifiers; the `encoding` and `compression` fields tell the verifier
 * which decoders to apply.
 *
 * The reference CLI (`tools/tf-packet/`) provides
 * sign/verify/inspect/fragment/reassemble/simulate-lora.
 */

import { deflateSync, inflateSync } from "node:zlib";
import { encode as cborEncode, decode as cborDecode } from "./cbor.js";
import { canonicalize } from "./canonical.js";
import { ed25519Sign, ed25519Verify, sha256 } from "./crypto.js";
import { isWithinWindow } from "./expiration.js";

export type Priority = "P0" | "P1" | "P2" | "P3" | "P4" | "P5";
export type PacketEncoding = "json" | "cbor";
export type Compression = "none" | "deflate";

export interface PacketFragmentHeader {
  fragment_id: string;
  index: number;
  count: number;
  total_payload_bytes: number;
  payload_digest: string;
}

import type { Packet } from "../generated/packet.js";

// Note: `Packet` is re-exported from generated/index.js — we intentionally
// only `import type` it here so this module doesn't double-export it.
export type { Packet };

export interface SignPacketArgs {
  packetId: string;
  source: string;
  destination: string;
  priority: Priority;
  payload: Uint8Array;
  encoding?: PacketEncoding;
  compression?: Compression;
  emergency?: boolean;
  expiresAt?: string;
  ttlHops?: number;
  routeConstraints?: string[];
  sessionRef?: string;
  privateKey: Uint8Array;
  signer: string;
  createdAt?: string;
}

export function packetSigningBytes(p: Packet): Uint8Array {
  const { signature: _signature, ...rest } = p;
  void _signature;
  return sha256(new TextEncoder().encode(canonicalize(rest as unknown)));
}

export async function signPacket(args: SignPacketArgs): Promise<Packet> {
  if (args.priority === "P0" && !args.emergency) {
    throw new Error("P0 priority is reserved for emergency packets");
  }
  const encoding: PacketEncoding = args.encoding ?? "cbor";
  const compression: Compression = args.compression ?? "none";
  const payloadBytes =
    encoding === "cbor"
      ? new Uint8Array(cborEncode({ raw: args.payload }))
      : new TextEncoder().encode(canonicalize({ raw: Buffer.from(args.payload).toString("base64") }));
  const finalBytes =
    compression === "deflate" ? new Uint8Array(deflateSync(payloadBytes)) : payloadBytes;
  const draft: Packet = {
    packet_version: "1",
    packet_id: args.packetId,
    source: args.source,
    destination: args.destination,
    priority: args.priority,
    created_at: args.createdAt ?? new Date().toISOString(),
    encoding,
    compression,
    payload: Buffer.from(finalBytes).toString("base64"),
    signature: { algorithm: "ed25519", signer: args.signer, signature: "" },
  };
  if (args.emergency) draft.emergency = true;
  if (args.expiresAt) draft.expires_at = args.expiresAt;
  if (args.ttlHops !== undefined) draft.ttl_hops = args.ttlHops;
  if (args.routeConstraints && args.routeConstraints.length > 0)
    draft.route_constraints = args.routeConstraints;
  if (args.sessionRef) draft.session_ref = args.sessionRef;
  const digest = packetSigningBytes(draft);
  const sig = await ed25519Sign(digest, args.privateKey);
  draft.signature = {
    algorithm: "ed25519",
    signer: args.signer,
    signature: Buffer.from(sig).toString("base64"),
  };
  return draft;
}

export interface VerifyPacketResult {
  ok: boolean;
  reason?: string;
  payload?: Uint8Array;
}

export async function verifyPacket(
  packet: Packet,
  publicKey: Uint8Array,
  now: string = new Date().toISOString(),
): Promise<VerifyPacketResult> {
  if (packet.packet_version !== "1") {
    return { ok: false, reason: `unsupported packet_version ${packet.packet_version}` };
  }
  if (packet.signature.signer !== packet.source) {
    return { ok: false, reason: "signature signer does not match source" };
  }
  if (packet.priority === "P0" && !packet.emergency) {
    return { ok: false, reason: "P0 reserved for emergency packets" };
  }
  if (packet.expires_at && !isWithinWindow({ valid_until: packet.expires_at }, now)) {
    return { ok: false, reason: "packet expired" };
  }
  const digest = packetSigningBytes(packet);
  const sigBytes = new Uint8Array(Buffer.from(packet.signature.signature, "base64"));
  const ok = await ed25519Verify(publicKey, digest, sigBytes);
  if (!ok) return { ok: false, reason: "signature verification failed" };
  // Decode payload back to bytes.
  const wireBytes = new Uint8Array(Buffer.from(packet.payload, "base64"));
  const decompressed =
    packet.compression === "deflate" ? new Uint8Array(inflateSync(wireBytes)) : wireBytes;
  let payload: Uint8Array;
  try {
    if (packet.encoding === "cbor") {
      const obj = cborDecode(decompressed) as { raw: Uint8Array };
      payload = obj.raw instanceof Uint8Array ? obj.raw : new Uint8Array(obj.raw as ArrayLike<number>);
    } else {
      const obj = JSON.parse(new TextDecoder().decode(decompressed)) as { raw: string };
      payload = new Uint8Array(Buffer.from(obj.raw, "base64"));
    }
  } catch (e) {
    return { ok: false, reason: `payload decode failed: ${(e as Error).message}` };
  }
  return { ok: true, payload };
}

/* -------------------------------------------------------------------------- */
/*  Fragmentation / reassembly                                                */
/* -------------------------------------------------------------------------- */

export interface FragmentOptions {
  /** Maximum payload bytes per fragment (the rest of the packet adds
   *  ~256 bytes of metadata). Default 256 (LoRa-friendly). */
  mtu?: number;
}

export async function fragmentPacket(
  source: Packet,
  privateKey: Uint8Array,
  opts: FragmentOptions = {},
): Promise<Packet[]> {
  const mtu = opts.mtu ?? 256;
  const fragmentId = `frag-${source.packet_id}`;
  const original = new Uint8Array(Buffer.from(source.payload, "base64"));
  const totalBytes = original.length;
  const digest = "sha256:" + Buffer.from(sha256(original)).toString("hex");
  if (totalBytes <= mtu) {
    return [source];
  }
  const count = Math.ceil(totalBytes / mtu);
  const out: Packet[] = [];
  for (let i = 0; i < count; i++) {
    const slice = original.slice(i * mtu, Math.min((i + 1) * mtu, totalBytes));
    const fragment: PacketFragmentHeader = {
      fragment_id: fragmentId,
      index: i,
      count,
      total_payload_bytes: totalBytes,
      payload_digest: digest,
    };
    const draft: Packet = {
      packet_version: "1",
      packet_id: `${source.packet_id}-${i}`,
      source: source.source,
      destination: source.destination,
      priority: source.priority,
      created_at: source.created_at,
      payload: Buffer.from(slice).toString("base64"),
      fragment,
      signature: { algorithm: "ed25519", signer: source.source, signature: "" },
    };
    if (source.emergency) draft.emergency = true;
    if (source.expires_at) draft.expires_at = source.expires_at;
    if (source.ttl_hops !== undefined) draft.ttl_hops = source.ttl_hops;
    if (source.route_constraints) draft.route_constraints = source.route_constraints;
    if (source.session_ref) draft.session_ref = source.session_ref;
    if (source.encoding) draft.encoding = source.encoding;
    if (source.compression) draft.compression = source.compression;
    const fragDigest = packetSigningBytes(draft);
    const sig = await ed25519Sign(fragDigest, privateKey);
    draft.signature = {
      algorithm: "ed25519",
      signer: source.source,
      signature: Buffer.from(sig).toString("base64"),
    };
    out.push(draft);
  }
  return out;
}

export interface ReassembleResult {
  ok: boolean;
  reason?: string;
  packetId?: string;
  payload?: Uint8Array;
}

export function reassembleFragments(fragments: Packet[]): ReassembleResult {
  if (fragments.length === 0) return { ok: false, reason: "no fragments" };
  const first = fragments[0]!.fragment;
  if (!first) return { ok: false, reason: "first fragment missing fragment header" };
  const fragId = first.fragment_id;
  const count = first.count;
  if (fragments.length !== count) {
    return { ok: false, reason: `expected ${count} fragments, got ${fragments.length}` };
  }
  const ordered: (Packet | undefined)[] = new Array(count).fill(undefined);
  for (const f of fragments) {
    if (!f.fragment || f.fragment.fragment_id !== fragId) {
      return { ok: false, reason: "mismatched fragment_id" };
    }
    if (f.fragment.count !== count) {
      return { ok: false, reason: "mismatched fragment count" };
    }
    if (ordered[f.fragment.index] !== undefined) {
      return { ok: false, reason: `duplicate fragment index ${f.fragment.index}` };
    }
    ordered[f.fragment.index] = f;
  }
  if (ordered.some((p) => p === undefined)) {
    return { ok: false, reason: "missing one or more fragment indices" };
  }
  const total = first.total_payload_bytes;
  const out = new Uint8Array(total);
  let offset = 0;
  for (const f of ordered as Packet[]) {
    const slice = new Uint8Array(Buffer.from(f.payload, "base64"));
    if (offset + slice.length > total) {
      return { ok: false, reason: "fragment overflows declared total bytes" };
    }
    out.set(slice, offset);
    offset += slice.length;
  }
  if (offset !== total) {
    return { ok: false, reason: `assembled ${offset} bytes, expected ${total}` };
  }
  const computed = "sha256:" + Buffer.from(sha256(out)).toString("hex");
  if (computed !== first.payload_digest) {
    return { ok: false, reason: "reassembled payload digest mismatch" };
  }
  return { ok: true, packetId: fragId, payload: out };
}

/* -------------------------------------------------------------------------- */
/*  LoRa-style transport simulator                                            */
/* -------------------------------------------------------------------------- */

export interface LoraSimOptions {
  /** Fraction of frames dropped, 0..1. Default 0. */
  packetLoss?: number;
  /** Bandwidth in bytes per second (used to model latency). Default 250 (SF7). */
  bandwidthBps?: number;
  /** Per-frame fixed latency in ms (RF + processing). Default 5000ms. */
  baseLatencyMs?: number;
  /** Optional seeded RNG for deterministic tests. */
  random?: () => number;
}

export interface SimulationResult {
  delivered: Packet[];
  dropped: Packet[];
  /** Cumulative simulated latency in ms. */
  totalLatencyMs: number;
}

/** Walk a list of packets through a one-way LoRa-style channel. Drops
 *  fragments per `packetLoss`, accumulates latency proportional to size
 *  divided by `bandwidthBps`. Pure simulation — no IO. */
export function simulateLora(packets: Packet[], opts: LoraSimOptions = {}): SimulationResult {
  const loss = opts.packetLoss ?? 0;
  const bw = opts.bandwidthBps ?? 250;
  const base = opts.baseLatencyMs ?? 5000;
  const rng = opts.random ?? Math.random;
  const delivered: Packet[] = [];
  const dropped: Packet[] = [];
  let totalLatencyMs = 0;
  for (const p of packets) {
    const sizeBytes = Buffer.from(canonicalize(p as unknown)).byteLength;
    const txMs = (sizeBytes / bw) * 1000;
    totalLatencyMs += base + txMs;
    if (rng() < loss) dropped.push(p);
    else delivered.push(p);
  }
  return { delivered, dropped, totalLatencyMs };
}

/* -------------------------------------------------------------------------- */
/*  Emergency authority + post-event quorum review                            */
/* -------------------------------------------------------------------------- */

export interface EmergencyInvocation {
  packet: Packet;
  /** The post-event quorum review that audits the emergency action.
   *  Required before evidence bundles can include the emergency event. */
  reviewBundleId?: string;
}

export function isEmergencyPacket(p: Packet): boolean {
  return p.priority === "P0" && p.emergency === true;
}

/** Build a packet bundle that pairs an emergency packet with its
 *  required post-event quorum review (TF-0011 \"emergency packets must
 *  be scoped, logged, and reviewable\"). */
export function emergencyReviewBundle(args: {
  bundleId: string;
  emergency: Packet;
  reviewPackets: Packet[];
  signer: string;
  privateKey: Uint8Array;
  transportHint?: "usb" | "qr-code" | "serial" | "lora" | "file-drop" | "manual";
  createdAt?: string;
}): Promise<{
  bundle_version: "1";
  bundle_id: string;
  label: string;
  packets: Packet[];
  transport_hint?: string;
  created_at: string;
  signature: { algorithm: string; signer: string; signature: string };
}> {
  if (!isEmergencyPacket(args.emergency)) {
    return Promise.reject(new Error("first packet must be a P0 emergency packet"));
  }
  if (args.reviewPackets.length === 0) {
    return Promise.reject(new Error("emergency review bundle requires at least one review packet"));
  }
  const draft = {
    bundle_version: "1" as const,
    bundle_id: args.bundleId,
    label: "emergency invocation + quorum review",
    packets: [args.emergency, ...args.reviewPackets],
    transport_hint: args.transportHint,
    created_at: args.createdAt ?? new Date().toISOString(),
    signature: {
      algorithm: "ed25519",
      signer: args.signer,
      signature: "",
    } as { algorithm: string; signer: string; signature: string },
  };
  const digest = sha256(
    new TextEncoder().encode(canonicalize({ ...draft, signature: undefined } as unknown)),
  );
  return ed25519Sign(digest, args.privateKey).then((sig) => {
    draft.signature.signature = Buffer.from(sig).toString("base64");
    return draft;
  });
}

File: crates/tf-types/src/packet.rs
#![allow(clippy::unusual_byte_groupings)]
//! Packet mode (TF-0011) — Rust mirror of
//! `tools/tf-types-ts/src/core/packet.ts`.

use crate::encoding::STANDARD;
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use flate2::read::DeflateDecoder;
use flate2::write::DeflateEncoder;
use flate2::Compression;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::io::{Read, Write};

use crate::canonicalize;
use crate::expiration::{is_within_window, Window};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct PacketSignatureEnvelope {
    pub algorithm: String,
    pub signer: String,
    pub signature: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct PacketFragmentHeader {
    pub fragment_id: String,
    pub index: u32,
    pub count: u32,
    pub total_payload_bytes: u32,
    pub payload_digest: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Packet {
    pub packet_version: String,
    pub packet_id: String,
    pub source: String,
    pub destination: String,
    pub priority: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub emergency: Option<bool>,
    pub created_at: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub expires_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub ttl_hops: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub route_constraints: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub encoding: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub compression: Option<String>,
    pub payload: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub session_ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub fragment: Option<PacketFragmentHeader>,
    pub signature: PacketSignatureEnvelope,
}

#[derive(Clone, Debug)]
pub struct SignPacketArgs<'a> {
    pub packet_id: String,
    pub source: String,
    pub destination: String,
    pub priority: String,
    pub payload: &'a [u8],
    pub encoding: Option<String>,
    pub compression: Option<String>,
    pub emergency: bool,
    pub expires_at: Option<String>,
    pub ttl_hops: Option<u32>,
    pub route_constraints: Option<Vec<String>>,
    pub session_ref: Option<String>,
    pub private_key: [u8; 32],
    pub signer: String,
    pub created_at: Option<String>,
}

pub fn packet_signing_bytes(p: &Packet) -> [u8; 32] {
    let mut value = serde_json::to_value(p).unwrap_or(Value::Null);
    if let Value::Object(map) = &mut value {
        map.remove("signature");
    }
    let canonical = canonicalize(&value).unwrap_or_default();
    Sha256::digest(canonical.as_bytes()).into()
}

pub fn sign_packet(args: SignPacketArgs<'_>) -> Result<Packet, String> {
    if args.priority == "P0" && !args.emergency {
        return Err("P0 priority is reserved for emergency packets".into());
    }
    let encoding = args.encoding.unwrap_or_else(|| "cbor".to_string());
    let compression = args.compression.unwrap_or_else(|| "none".to_string());

    // Wrap payload in canonical envelope before encoding.
    let payload_bytes: Vec<u8> = if encoding == "cbor" {
        serde_cbor_envelope(args.payload)
    } else {
        let canonical = canonicalize(&serde_json::json!({
            "raw": STANDARD.encode(args.payload),
        }))
        .map_err(|e| format!("canonicalize: {}", e))?;
        canonical.into_bytes()
    };

    let final_bytes = if compression == "deflate" {
        let mut enc = DeflateEncoder::new(Vec::new(), Compression::default());
        enc.write_all(&payload_bytes).map_err(|e| e.to_string())?;
        enc.finish().map_err(|e| e.to_string())?
    } else {
        payload_bytes
    };

    let mut draft = Packet {
        packet_version: "1".into(),
        packet_id: args.packet_id,
        source: args.source.clone(),
        destination: args.destination,
        priority: args.priority,
        emergency: if args.emergency { Some(true) } else { None },
        created_at: args.created_at.unwrap_or_else(now_iso8601),
        expires_at: args.expires_at,
        ttl_hops: args.ttl_hops,
        route_constraints: args.route_constraints.filter(|r| !r.is_empty()),
        encoding: Some(encoding),
        compression: Some(compression),
        payload: STANDARD.encode(&final_bytes),
        session_ref: args.session_ref,
        fragment: None,
        signature: PacketSignatureEnvelope {
            algorithm: "ed25519".into(),
            signer: args.signer.clone(),
            signature: String::new(),
        },
    };
    let digest = packet_signing_bytes(&draft);
    let signing = SigningKey::from_bytes(&args.private_key);
    let sig: Signature = signing.sign(&digest);
    draft.signature.signature = STANDARD.encode(sig.to_bytes());
    Ok(draft)
}

#[derive(Debug)]
pub struct VerifyPacketResult {
    pub ok: bool,
    pub reason: Option<String>,
    pub payload: Option<Vec<u8>>,
}

pub fn verify_packet(packet: &Packet, public_key: &[u8; 32], now: &str) -> VerifyPacketResult {
    let rejected = |r: &str| VerifyPacketResult {
        ok: false,
        reason: Some(r.to_string()),
        payload: None,
    };
    if packet.packet_version != "1" {
        return rejected(&format!(
            "unsupported packet_version {}",
            packet.packet_version
        ));
    }
    if packet.signature.signer != packet.source {
        return rejected("signature signer does not match source");
    }
    if packet.priority == "P0" && packet.emergency != Some(true) {
        return rejected("P0 reserved for emergency packets");
    }
    if let Some(expires) = &packet.expires_at {
        let window = Window {
            valid_until: Some(expires.as_str()),
            ..Window::default()
        };
        if !is_within_window(&window, now) {
            return rejected("packet expired");
        }
    }
    let digest = packet_signing_bytes(packet);
    let sig_bytes = match STANDARD.decode(&packet.signature.signature) {
        Ok(b) => b,
        Err(e) => return rejected(&format!("signature base64 decode: {}", e)),
    };
    let sig = match Signature::from_slice(&sig_bytes) {
        Ok(s) => s,
        Err(e) => return rejected(&format!("signature parse: {}", e)),
    };
    let vk = match VerifyingKey::from_bytes(public_key) {
        Ok(v) => v,
        Err(e) => return rejected(&format!("verifying key: {}", e)),
    };
    if vk.verify(&digest, &sig).is_err() {
        return rejected("signature verification failed");
    }
    let wire = match STANDARD.decode(&packet.payload) {
        Ok(b) => b,
        Err(e) => return rejected(&format!("payload base64: {}", e)),
    };
    let decompressed = if packet.compression.as_deref() == Some("deflate") {
        let mut dec = DeflateDecoder::new(&wire[..]);
        let mut out = Vec::new();
        if dec.read_to_end(&mut out).is_err() {
            return rejected("deflate inflate failed");
        }
        out
    } else {
        wire
    };
    let payload = match packet.encoding.as_deref() {
        Some("json") | None => match serde_json::from_slice::<Value>(&decompressed) {
            Ok(v) => match v.get("raw").and_then(|r| r.as_str()) {
                Some(b64) => match STANDARD.decode(b64) {
                    Ok(b) => b,
                    Err(_) => return rejected("payload base64 inner"),
                },
                None => return rejected("json envelope missing raw"),
            },
            Err(e) => return rejected(&format!("json decode: {}", e)),
        },
        Some("cbor") => match decode_cbor_envelope(&decompressed) {
            Ok(b) => b,
            Err(e) => return rejected(&format!("cbor decode: {}", e)),
        },
        Some(other) => return rejected(&format!("unknown encoding {}", other)),
    };
    VerifyPacketResult {
        ok: true,
        reason: None,
        payload: Some(payload),
    }
}

#[derive(Clone, Debug, Default)]
pub struct FragmentOptions {
    pub mtu: Option<usize>,
}

pub fn fragment_packet(
    source: &Packet,
    private_key: &[u8; 32],
    opts: FragmentOptions,
) -> Vec<Packet> {
    let mtu = opts.mtu.unwrap_or(256);
    let original = STANDARD.decode(&source.payload).unwrap_or_default();
    let total_bytes = original.len();
    if total_bytes <= mtu {
        return vec![source.clone()];
    }
    let count = total_bytes.div_ceil(mtu);
    let digest_hex = sha256_hex(&original);
    let payload_digest = format!("sha256:{}", digest_hex);
    let fragment_id = format!("frag-{}", source.packet_id);
    let mut out = Vec::with_capacity(count);
    for i in 0..count {
        let start = i * mtu;
        let end = (start + mtu).min(total_bytes);
        let slice = &original[start..end];
        let mut draft = Packet {
            packet_version: "1".into(),
            packet_id: format!("{}-{}", source.packet_id, i),
            source: source.source.clone(),
            destination: source.destination.clone(),
            priority: source.priority.clone(),
            emergency: source.emergency,
            created_at: source.created_at.clone(),
            expires_at: source.expires_at.clone(),
            ttl_hops: source.ttl_hops,
            route_constraints: source.route_constraints.clone(),
            encoding: source.encoding.clone(),
            compression: source.compression.clone(),
            payload: STANDARD.encode(slice),
            session_ref: source.session_ref.clone(),
            fragment: Some(PacketFragmentHeader {
                fragment_id: fragment_id.clone(),
                index: i as u32,
                count: count as u32,
                total_payload_bytes: total_bytes as u32,
                payload_digest: payload_digest.clone(),
            }),
            signature: PacketSignatureEnvelope {
                algorithm: "ed25519".into(),
                signer: source.source.clone(),
                signature: String::new(),
            },
        };
        let digest = packet_signing_bytes(&draft);
        let signing = SigningKey::from_bytes(private_key);
        let sig: Signature = signing.sign(&digest);
        draft.signature.signature = STANDARD.encode(sig.to_bytes());
        out.push(draft);
    }
    out
}

#[derive(Debug)]
pub struct ReassembleResult {
    pub ok: bool,
    pub reason: Option<String>,
    pub packet_id: Option<String>,
    pub payload: Option<Vec<u8>>,
}

pub fn reassemble_fragments(fragments: &[Packet]) -> ReassembleResult {
    let rejected = |r: &str| ReassembleResult {
        ok: false,
        reason: Some(r.to_string()),
        packet_id: None,
        payload: None,
    };
    if fragments.is_empty() {
        return rejected("no fragments");
    }
    let header = match fragments[0].fragment.as_ref() {
        Some(h) => h.clone(),
        None => return rejected("first fragment missing fragment header"),
    };
    if fragments.len() != header.count as usize {
        return rejected(&format!(
            "expected {} fragments, got {}",
            header.count,
            fragments.len()
        ));
    }
    let mut ordered: Vec<Option<&Packet>> = vec![None; header.count as usize];
    for f in fragments {
        let h = match f.fragment.as_ref() {
            Some(h) => h,
            None => return rejected("fragment missing header"),
        };
        if h.fragment_id != header.fragment_id {
            return rejected("mismatched fragment_id");
        }
        if h.count != header.count {
            return rejected("mismatched fragment count");
        }
        if ordered[h.index as usize].is_some() {
            return rejected(&format!("duplicate fragment index {}", h.index));
        }
        ordered[h.index as usize] = Some(f);
    }
    let mut out = Vec::with_capacity(header.total_payload_bytes as usize);
    for slot in ordered {
        match slot {
            Some(p) => {
                let bytes = match STANDARD.decode(&p.payload) {
                    Ok(b) => b,
                    Err(e) => return rejected(&format!("base64: {}", e)),
                };
                out.extend_from_slice(&bytes);
            }
            None => return rejected("missing fragment slot"),
        }
    }
    if out.len() != header.total_payload_bytes as usize {
        return rejected(&format!(
            "assembled {} bytes, expected {}",
            out.len(),
            header.total_payload_bytes
        ));
    }
    let computed = format!("sha256:{}", sha256_hex(&out));
    if computed != header.payload_digest {
        return rejected("reassembled payload digest mismatch");
    }
    ReassembleResult {
        ok: true,
        reason: None,
        packet_id: Some(header.fragment_id),
        payload: Some(out),
    }
}

fn serde_cbor_envelope(raw: &[u8]) -> Vec<u8> {
    let value = crate::cbor::Value::Map(vec![(
        crate::cbor::Value::Text("raw".into()),
        crate::cbor::Value::Bytes(raw.to_vec()),
    )]);
    crate::cbor::encode(&value).expect("cbor encode")
}

fn decode_cbor_envelope(bytes: &[u8]) -> Result<Vec<u8>, String> {
    let value = crate::cbor::decode(bytes).map_err(|e| format!("cbor: {}", e))?;
    match value {
        crate::cbor::Value::Map(entries) => {
            for (k, v) in entries {
                if matches!(k, crate::cbor::Value::Text(ref s) if s == "raw") {
                    if let crate::cbor::Value::Bytes(b) = v {
                        return Ok(b);
                    }
                }
            }
            Err("cbor envelope missing raw".into())
        }
        _ => Err("cbor envelope not a map".into()),
    }
}

fn sha256_hex(b: &[u8]) -> String {
    let digest = Sha256::digest(b);
    digest.iter().map(|byte| format!("{:02x}", byte)).collect()
}

fn now_iso8601() -> String {
    let secs = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64;
    let (y, m, d, h, mi, s) = secs_to_ymdhms(secs);
    format!("{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z", y, m, d, h, mi, s)
}

fn secs_to_ymdhms(secs: i64) -> (i32, u32, u32, u32, u32, u32) {
    let days = secs.div_euclid(86_400);
    let time = secs.rem_euclid(86_400);
    let hour = (time / 3600) as u32;
    let minute = ((time % 3600) / 60) as u32;
    let second = (time % 60) as u32;
    let z = days + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let m = if mp < 10 {
        (mp + 3) as u32
    } else {
        (mp - 9) as u32
    };
    let year = if m <= 2 { y + 1 } else { y };
    (year as i32, m, d, hour, minute, second)
}

/* ----------------------------------------------------------------------- */
/*  LoRa-style channel simulation (mirror of TS `simulateLora`)            */
/* ----------------------------------------------------------------------- */

#[derive(Clone, Debug, Default)]
pub struct LoraSimOptions {
    /// Per-packet drop probability ∈ [0, 1]. Default 0 (lossless).
    pub packet_loss: Option<f64>,
    /// Bandwidth in bytes/sec. Default 250.
    pub bandwidth_bps: Option<f64>,
    /// Base latency in ms. Default 5000.
    pub base_latency_ms: Option<f64>,
}

#[derive(Debug, Default)]
pub struct LoraSimResult {
    pub delivered: Vec<Packet>,
    pub dropped: Vec<Packet>,
    /// Cumulative simulated latency, ms.
    pub total_latency_ms: f64,
}

/// Walk a list of packets through a one-way LoRa-style channel. Drops
/// packets per `packet_loss`, accumulates latency proportional to size /
/// `bandwidth_bps`. Pure simulation — no IO. The optional `rng_seed`
/// argument makes the result deterministic for tests.
pub fn simulate_lora(
    packets: &[Packet],
    opts: LoraSimOptions,
    rng_seed: Option<u64>,
) -> LoraSimResult {
    let loss = opts.packet_loss.unwrap_or(0.0);
    let bw = opts.bandwidth_bps.unwrap_or(250.0);
    let base = opts.base_latency_ms.unwrap_or(5000.0);
    let mut state = rng_seed.unwrap_or(0xdeadbeef_dead_beefu64);
    let mut next = move || {
        // xorshift64* — small deterministic PRNG, values in (0,1).
        state ^= state << 13;
        state ^= state >> 7;
        state ^= state << 17;

        (state.wrapping_mul(0x2545_F491_4F6C_DD1Du64) >> 11) as f64 / (1u64 << 53) as f64
    };
    let mut delivered = Vec::with_capacity(packets.len());
    let mut dropped: Vec<Packet> = Vec::new();
    let mut total_latency_ms = 0.0_f64;
    for p in packets {
        let canonical = serde_json::to_string(p).unwrap_or_default();
        let size_bytes = canonical.len() as f64;
        let tx_ms = (size_bytes / bw) * 1000.0;
        total_latency_ms += base + tx_ms;
        if next() < loss {
            dropped.push(p.clone());
        } else {
            delivered.push(p.clone());
        }
    }
    LoraSimResult {
        delivered,
        dropped,
        total_latency_ms,
    }
}

#[cfg(test)]
mod lora_tests {
    use super::*;
    use rand::rngs::OsRng;

    fn fixture(packet_id: &str) -> Packet {
        let signer = "tf:actor:agent:example.com/x";
        let mut signer_seed = [0u8; 32];
        rand::RngCore::fill_bytes(&mut OsRng, &mut signer_seed);
        sign_packet(SignPacketArgs {
            packet_id: packet_id.into(),
            source: signer.into(),
            destination: "tf:actor:service:example.com/d".into(),
            priority: "P3".into(),
            payload: b"hi",
            encoding: None,
            compression: None,
            emergency: false,
            expires_at: None,
            ttl_hops: None,
            route_constraints: None,
            session_ref: None,
            private_key: signer_seed,
            signer: signer.into(),
            created_at: Some("2026-04-24T12:00:00Z".into()),
        })
        .expect("sign")
    }

    #[test]
    fn lossless_channel_delivers_everything() {
        let packets = vec![fixture("a"), fixture("b"), fixture("c")];
        let r = simulate_lora(
            &packets,
            LoraSimOptions {
                packet_loss: Some(0.0),
                bandwidth_bps: Some(250.0),
                base_latency_ms: Some(5000.0),
            },
            Some(1),
        );
        assert_eq!(r.delivered.len(), 3);
        assert_eq!(r.dropped.len(), 0);
        assert!(r.total_latency_ms > 15_000.0);
    }

    #[test]
    fn full_loss_drops_everything() {
        let packets = vec![fixture("a"), fixture("b")];
        let r = simulate_lora(
            &packets,
            LoraSimOptions {
                packet_loss: Some(1.0),
                ..Default::default()
            },
            Some(42),
        );
        assert_eq!(r.delivered.len(), 0);
        assert_eq!(r.dropped.len(), 2);
    }

    #[test]
    fn deterministic_with_seed() {
        let packets: Vec<Packet> = (0..10).map(|i| fixture(&format!("pkt-{}", i))).collect();
        let r1 = simulate_lora(
            &packets,
            LoraSimOptions {
                packet_loss: Some(0.5),
                ..Default::default()
            },
            Some(99),
        );
        let r2 = simulate_lora(
            &packets,
            LoraSimOptions {
                packet_loss: Some(0.5),
                ..Default::default()
            },
            Some(99),
        );
        assert_eq!(r1.delivered.len(), r2.delivered.len());
        assert_eq!(r1.dropped.len(), r2.dropped.len());
    }
}
